### PR TITLE
Solve bug in IDP planner

### LINKED ISSUE #10053: Queries That Used To Work in Previous Versions Now Generate Errors
I have detected two queries that used to work in previous versions of Neo4j (confirmed working in 3.0) which now generate errors.

You can try the queries on any graph (empty or with data).

- Neo4j version: 3.2.2 and 3.2.3 (confirmed)
- Operating system: Windows and the OS of Neo4j sandbox
- API/Driver: HTTP, Native (C#), Neo4jClient (which uses HTTP).

## Steps to reproduce

1. Execute any of the example queries in either the Neo4j browser shell, or with any native client (I tested with C#), or Neo4jClient.

Example Query 1:
```
MATCH (a:A)
WITH a
WHERE a.prop
MATCH (c:C)
WHERE ( NOT (a)-[:P]->() OR (a)-[:P]->(c) )
WITH a, c
MATCH (a)-[:Q]->()-[:R]->(b:B)
RETURN a, c, b
```

Example Query 2:
```
MATCH (ts)
MATCH (k)-[:M]->(sta)
OPTIONAL MATCH (sta)<-[:N]-(p)
WITH k, ts, coalesce(p, sta) AS ab
MATCH (d:A)
WHERE d.Id = ab.OtherId
WITH k, ts, d
MATCH (ts)-[:R]->(f)
RETURN k, ts, f, d
```

## Expected results

If you have an empty graph or any graph that do not match the rules you should get an empty result set

## Actual results

For the first query you get the following error, for the others you get a very similar error:

```
Expected 
RegularPlannerQuery(QueryGraph(Set(),Set(IdName(a)),Set(),Selections(Set(Predicate(Set(IdName(a)),HasLabels(Variable(a),List(LabelName(A)))))),Vector(),Set(),Set(),List()),RegularQueryProjection(Map(a -> Variable(a),   FRESHID27 -> Property(Variable(a),PropertyKeyName(prop))),QueryShuffle(List(),None,None)),Some(RegularPlannerQuery(QueryGraph(Set(PatternRelationship(IdName(  UNNAMED115),(IdName(a),IdName(  UNNAMED122)),OUTGOING,List(RelTypeName(Q)),SimplePatternLength), PatternRelationship(IdName(  UNNAMED124),(IdName(  UNNAMED122),IdName(b)),OUTGOING,List(RelTypeName(R)),SimplePatternLength)),Set(IdName(c), IdName(a), IdName(  UNNAMED122), IdName(b)),Set(IdName(a), IdName(  FRESHID27)),Selections(Set(Predicate(Set(IdName(b)),HasLabels(Variable(b),List(LabelName(B)))), Predicate(Set(IdName(a), IdName(c)),Ors(Set(LessThanOrEqual(GetDegree(Variable(a),Some(RelTypeName(P)),OUTGOING),SignedDecimalIntegerLiteral(0)), PatternExpression(RelationshipsPattern(RelationshipChain(NodePattern(Some(Variable(a)),List(),None),RelationshipPattern(None,List(RelTypeName(P)),None,None,OUTGOING,false),NodePattern(Some(Variable(c)),List(),None))))))), Predicate(Set(IdName(c)),HasLabels(Variable(c),List(LabelName(C)))), Predicate(Set(IdName(  FRESHID27)),Variable(  FRESHID27)))),Vector(),Set(),Set(),List()),RegularQueryProjection(Map(a -> Variable(a), c -> Variable(c), b -> Variable(b)),QueryShuffle(List(),None,None)),None))) 


Instead, got: 
RegularPlannerQuery(QueryGraph(Set(),Set(IdName(a)),Set(),Selections(Set(Predicate(Set(IdName(a)),HasLabels(Variable(a),List(LabelName(A)))))),Vector(),Set(),Set(),List()),RegularQueryProjection(Map(a -> Variable(a),   FRESHID27 -> Property(Variable(a),PropertyKeyName(prop))),QueryShuffle(List(),None,None)),Some(RegularPlannerQuery(QueryGraph(Set(PatternRelationship(IdName(  UNNAMED115),(IdName(a),IdName(  UNNAMED122)),OUTGOING,List(RelTypeName(Q)),SimplePatternLength), PatternRelationship(IdName(  UNNAMED124),(IdName(  UNNAMED122),IdName(b)),OUTGOING,List(RelTypeName(R)),SimplePatternLength)),Set(IdName(a), IdName(  UNNAMED122), IdName(b)),Set(IdName(a), IdName(  FRESHID27)),Selections(Set(Predicate(Set(IdName(  FRESHID27)),Variable(  FRESHID27)), Predicate(Set(IdName(b)),HasLabels(Variable(b),List(LabelName(B)))))),Vector(),Set(),Set(),List()),RegularQueryProjection(Map(a -> Variable(a), c -> Variable(c), b -> Variable(b)),QueryShuffle(List(),None,None)),None)))
```

You can actually "fix" the queries by changing them a little.

Fix for Query 1 (notice the removal of the first WITH):
```
MATCH (a:A)
WHERE a.prop
MATCH (c:C)
WHERE ( NOT (a)-[:P]->() OR (a)-[:P]->(c) )
WITH a, c
MATCH (a)-[:Q]->()-[:R]->(b:B)
RETURN a, c, b
```

Fix for Query 2 (notice that instead of using a MATCH after the coalesce, it uses an OPTIONAL MATCH):
```
MATCH (ts)
MATCH (k)-[:M]->(sta)
OPTIONAL MATCH (sta)<-[:N]-(p)
WITH k, ts, coalesce(sta, p) AS sta
OPTIONAL MATCH (d:A)
WHERE d.Id = sta.OtherId
WITH k, ts, d
MATCH (ts)-[:R]->(f)
RETURN k, ts, f, d
```

### PR DESCRIPTION
**Pull request by systay**
In rare situations, the cost planner was failing to solve all parts
MATCH pattern.

Fixes #10053

### CODE REVIEW

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala [hunk 1]
```diff
@@ -29,6 +29,28 @@ import scala.collection.mutable.ArrayBuffer
...
+    val result = executeWithAllPlannersAndCompatibilityMode(
+      """MATCH (a)
+        |WITH a WHERE true
+        |MATCH (c), (a)-[r]->(x)
+        |WHERE (a)-->(c)
+        |RETURN *""".stripMargin)
+
+    result.size should equal(0)// does not throw
+  }
+
```
Comment (sherfert) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: Is this query also fixed?

```
MATCH (ts)
MATCH (k)-[:M]->(sta)
OPTIONAL MATCH (sta)<-[:N]-(p)
WITH k, ts, coalesce(p, sta) AS ab
MATCH (d:A)
WHERE d.Id = ab.OtherId
WITH k, ts, d
MATCH (ts)-[:R]->(f)
RETURN k, ts, f, d
```
If yes, can you also add a test here?

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/logical/idp/SingleComponentPlanner.scala [hunk 2]
```diff
@@ -92,7 +92,16 @@ case class SingleComponentPlanner(monitor: IDPQueryGraphSolverMonitor,
     for (pattern <- qg.patternRelationships)
       yield {
         val accessPlans = planSinglePattern(qg, pattern, leaves).map(plan => kit.select(plan, qg))
-        val bestAccessor = kit.pickBest(accessPlans).getOrElse(throw new InternalException("Found no access plan for a pattern relationship in a connected component. This must not happen."))
+        val relationships = qg.patternRelationships.map(_.name)
+        val argumentContainsRelationship = qg.argumentIds.exists(relationships.contains)
+        val plans = if (argumentContainsRelationship) {
+          accessPlans
+        } else {
+          val maxSize = accessPlans.map(_.availableSymbols.size).max
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/logical/idp/SingleComponentPlanner.scala]: Future me - and to some extent present me - will be confused why we do this. Can we have a short comment explaining the rationale behind these lines of code.

**Review by pontusmelke — CHANGES_REQUESTED** [review]
Nice, just had a minor thing and Satia had a good question. Once that is fixed I'll merge!

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala (hunk 1 above)
Comment (OliviaYtterbrink) [reply enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: There is a test added for this.

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/logical/idp/SingleComponentPlanner.scala (hunk 2 above)
Comment (systay) [reply community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/logical/idp/SingleComponentPlanner.scala]: Thanks for making me think! When I revisited this, it was clear that this was a hack and not the real solution. I take it back! Look in QueryGraph.connectedComponents for the real solution

**Comment by systay** [comment]
@sherfert's concern was solved with @OliviaYtterbrink's commit

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/logical/idp/SingleComponentPlanner.scala [hunk 3]
```diff
@@ -91,8 +91,11 @@ case class SingleComponentPlanner(monitor: IDPQueryGraphSolverMonitor,
   private def initTable(qg: QueryGraph, kit: QueryPlannerKit, leaves: Set[LogicalPlan])(implicit context: LogicalPlanningContext) = {
     for (pattern <- qg.patternRelationships)
       yield {
-        val accessPlans = planSinglePattern(qg, pattern, leaves).map(plan => kit.select(plan, qg))
-        val bestAccessor = kit.pickBest(accessPlans).getOrElse(throw new InternalException("Found no access plan for a pattern relationship in a connected component. This must not happen."))
+        val relationships = qg.patternRelationships.map(_.name)
+        val argumentContainsRelationship = qg.argumentIds.exists(relationships.contains)
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/logical/idp/SingleComponentPlanner.scala]: not used?

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/logical/idp/SingleComponentPlanner.scala [hunk 4]
```diff
@@ -91,8 +91,11 @@ case class SingleComponentPlanner(monitor: IDPQueryGraphSolverMonitor,
   private def initTable(qg: QueryGraph, kit: QueryPlannerKit, leaves: Set[LogicalPlan])(implicit context: LogicalPlanningContext) = {
     for (pattern <- qg.patternRelationships)
       yield {
-        val accessPlans = planSinglePattern(qg, pattern, leaves).map(plan => kit.select(plan, qg))
-        val bestAccessor = kit.pickBest(accessPlans).getOrElse(throw new InternalException("Found no access plan for a pattern relationship in a connected component. This must not happen."))
+        val relationships = qg.patternRelationships.map(_.name)
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/logical/idp/SingleComponentPlanner.scala]: and this one is not really used either

File: enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala [hunk 5]
```diff
@@ -42,11 +42,11 @@ class MatchAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisticsTe
...
-        |MATCH (c), (a)-[r]->(x)
-        |WHERE (a)-->(c)
-        |RETURN *""".stripMargin)
+    """MATCH (a)
+      |WITH a WHERE true
+      |MATCH (c), (a)-[r]->(x)
+      |WHERE a.foo = c.bar
+      |RETURN *""".stripMargin)
 
     result.size should equal(0)// does not throw
```
Comment (pontusmelke) [inline enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/MatchAcceptanceTest.scala]: Is it a lot of extra work to make the test return at least one row and then assert on the result? Maybe add to MatchAcceptance.feature instead ( or maybe it is not there in 3.2?).

**Comment by pontusmelke** [comment]
@systay Still some minor comment, and now there seems to be an actual test failure.

**Review by pontusmelke — APPROVED** [review]

**Comment by pontusmelke** [comment]
@systay still some failing tests

**Pull request merged**